libs/blockscout-auth/src/auth.rs

```diff
@@ -0,0 +1,347 @@
+use cookie::Cookie;
+use reqwest::{
+    header::{HeaderMap, HeaderValue},
+    StatusCode,
+};
+use serde::Deserialize;
+use std::collections::HashMap;
+use thiserror::Error;
+use tonic::{codegen::http::header::COOKIE, metadata::MetadataMap};
+use url::Url;
+
+const JWT_TOKEN_NAME: &str = "_explorer_key";
+const CSRF_TOKEN_NAME: &str = "x-csrf-token";
+const API_KEY_NAME: &str = "api_key";
+
+#[derive(Debug)]
+pub struct AuthSuccess {
+    pub user_id: String,
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(untagged)]
+enum BlockscoutAuthResponse {
+    Success(BlockscoutAuthSuccessResponse),
+    Error(BlockscoutAuthErrorResponse),
+}
+
+#[derive(Debug, Clone, Deserialize)]
+struct BlockscoutAuthErrorResponse {
+    message: String,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Deserialize)]
+struct BlockscoutAuthSuccessResponse {
+    avatar: String,
+    email: String,
+    id: i64,
+    name: String,
+    nickname: String,
+    uid: String,
+    watchlist_id: i64,
+}
+
+impl From<BlockscoutAuthSuccessResponse> for AuthSuccess {
+    fn from(value: BlockscoutAuthSuccessResponse) -> Self {
+        Self {
+            user_id: value.id.to_string(),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("invalid jwt token: {0}")]
+    InvalidJwt(String),
+    #[error("invalid csrf token: {0}")]
+    InvalidCsrfToken(String),
+    #[error("user is unauthorized: {0}")]
+    Unauthorized(String),
+    #[error("blockscout invalid response: {0}")]
+    BlockscoutApi(String),
+    #[error("internal error: {0}")]
+    InternalError(String),
+}
+
+pub async fn auth_from_metadata(
+    metadata: &MetadataMap,
+    is_safe_http_method: bool,
+    blockscout_host: &Url,
+    blockscout_api_key: Option<&str>,
+) -> Result<AuthSuccess, Error> {
+    let jwt = extract_jwt(metadata)?;
+    let csrf_token = if is_safe_http_method {
+        None
+    } else {
+        Some(extract_csrf_token(metadata)?)
+    };
+    auth_from_tokens(&jwt, csrf_token, blockscout_host, blockscout_api_key).await
+}
+
+pub async fn auth_from_tokens(
+    jwt: &str,
+    csrf_token: Option<&str>,
+    blockscout_host: &Url,
+    blockscout_api_key: Option<&str>,
+) -> Result<AuthSuccess, Error> {
```

Comment:
Why not to return `BlockscoutAuthSuccessResponse` (just renamed to `AuthSuccess`)?

Reply:
the problem here is that blockscout returns `id: i64`, but in admin-rs we always need String type, so I decided to introduce new struct AuthSuccess to separate blockscout's response and library's response

Reply:
I suggest for the lib to return the same output as blockscout does (including `id` as `i64` if that is the case). I believe it should be the responsibility of the client application to decide what fields are needed and how they should be interpreted. It is better not to make those a library concerns

Reply:
ok